src/io/port.rs

```diff
@@ -0,0 +1,94 @@
+//! Solaris/Illumos event ports.
+
+use crate::backend::{c, io::syscalls};
+use crate::fd::{AsFd, AsRawFd, OwnedFd};
+use crate::io;
+
+use super::PollFlags;
+
+use core::time::Duration;
+
+/// The structure representing a port event.
+#[repr(transparent)]
+pub struct Event(pub(crate) c::port_event);
+
+impl Event {
+    /// Get the events associated with this event.
+    pub fn events(&self) -> i32 {
+        self.0.portev_events
+    }
+
+    /// Get the event source associated with this event.
+    pub fn object(&self) -> usize {
+        self.0.portev_object
+    }
+
+    /// Get the userdata associated with this event.
+    pub fn userdata(&self) -> *mut () {
+        self.0.portev_user.cast()
+    }
+}
+
+/// `port_create()`- Creates a new port.
+pub fn port_create() -> io::Result<OwnedFd> {
+    syscalls::port_create()
+}
+
+/// `port_associate()`- Associates a file descriptor with a port.
+pub fn port_associate_fd(
+    port: &impl AsFd,
+    object: &impl AsFd,
```

Comment:
Do you know what the behavior is if object is closed while it's associated with a port?

Also, is it possible to get the fd back from the port after this call?

Reply:
> Do you know what the behavior is if object is closed while it's associated with a port?

Hmm, going through the man pages for OpenSolaris and illumos, I can't find any documentation for what happens. I'm assuming it's the same for `epoll` and `kqueue`, but you know that they say about assumptions.

> Also, is it possible to get the fd back from the port after this call?

The `object` method of the `Event` structure contains the file descriptor. This can't be explicitly the FD because for `EVENT_SOURCE_FILE` it may also be a pointer to a structure describing the file.

Reply:
If it's the file descriptor, then I'm nervous about `AsFd`/`BorrowedFd` in the same I am about them for `kqueue`. It's important to me to be rigorous about borrows, even if there's no actual use-after-close hazard, because I'd like to eventually have things like I/O safety checking in miri. As in `kqueue`, what would you think about using `RawFd` and unsafe here?

The thing that hopefully makes this ok is that most users don't use these APIs directly; they're most often used in async runtimes and similar, which have to be carefully written, but then they can offer a higher-level abstraction to their users where it's easier to have safety invariants.

Reply:
`polling` uses `RawFd` and `unsafe` for now, and getting it to the point where it wouldn't would require some serious chicanery. I'll edit this PR and the `kqueue` one to use `RawFd`.